usr/opt/OSCX200/src/demos/xgas/xgas.man
.\"
.\" *****************************************************************
.\" *                                                               *
.\" *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
.\" *                                                               *
.\" *   All Rights Reserved.  Unpublished rights  reserved  under   *
.\" *   the copyright laws of the United States.                    *
.\" *                                                               *
.\" *   The software contained on this media  is  proprietary  to   *
.\" *   and  embodies  the  confidential  technology  of  Digital   *
.\" *   Equipment Corporation.  Possession, use,  duplication  or   *
.\" *   dissemination of the software and media is authorized only  *
.\" *   pursuant to a valid written license from Digital Equipment  *
.\" *   Corporation.                                                *
.\" *                                                               *
.\" *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
.\" *   by the U.S. Government is subject to restrictions  as  set  *
.\" *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
.\" *   or  in  FAR 52.227-19, as applicable.                       *
.\" *                                                               *
.\" *****************************************************************
.\"
.\"
.\" HISTORY
.\"
.TH XGAS 1 "Release 5" "X Version 11"
.SH NAME
xgas - animated simulation of an ideal gas
.SH SYNOPSIS
.B "xgas"
[-option ...]
.SH DESCRIPTION
.PP
\fIxgas\fP
is a physical simulation of an ideal gas in a heated box.
Gas molecules move around the box with velocities
dependent on their temperature.
A chamber consisting of two boxes contains the gas molecules;
the temperature of each box can be independently controlled
by a scrollbar.
When gas molecules collide with the walls, their temperature
approaches that of the box.
.PP
Use mouse button 1 to create molecules one at a time
at the cursor position.
Use mouse button 2 to create the maximum number of molecules
at the cursor position.
.SH OPTIONS
.TP 8
This is a toolkit program;
all the usual toolkit options are available.
.SH "X DEFAULTS"
The \fIxgas\fP program uses the following X resources:
.TP 8
.B timeStepSize
Specifies the simulated time duration in microseconds for each cycle of
computation.
.TP 8
.B delay
Specifies the real time interval between timestep computations.
.TP 8
.B randomBounce
In each wall collision, a molecule bounces elastically
(angle of incidence = angle of reflection).
A component of randomness is added to this angle.
RandomBounce varies from 0.0 (no randomness)
to 1.0 (completely random angle of incidence).
.TP 8
.B equilibrium
During each wall collision,
a molecule's kinetic energy approaches that
corresponding to the temperature of the wall.
If equilibrium is 1.0,
the molecule reaches the wall temperature immediately.
For values between 1.0 and 0.0,
the molecule approaches the temperature of the wall more slowly.
.TP 8
.B maxMolecules
Specifies the maximum number of molecules in the simulation.
.SH "SEE ALSO"
X(1), xwd(1)
.SH BUGS
.PP
When the chamber is resized,
molecules should be rearranged appropriately.
Instead,
the molecule arrays are reinitialized.
.SH COPYRIGHT
Copyright 1991, Massachusetts Institute of Technology.
.SH AUTHOR
Larry Medwin


usr/opt/OSCX200/src/demos/xgas/XGas.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
/* $XConsortium: XGas.c,v 1.1 91/04/18 09:48:25 dave Exp $ */

/* Copyright	Massachusetts Institute of Technology	1987, 1988
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of M.I.T. not be used in advertising or
 * publicity pertaining to distribution of the software without specific,
 * written prior permission.  M.I.T. makes no representations about the
 * suitability of this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 *
 * M.I.T. DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING ALL
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO EVENT SHALL M.I.T.
 * BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
 * OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 */

#include <X11/IntrinsicP.h>
#include <X11/StringDefs.h>
#include "XGasP.h"

static XtResource resources[] = {
#define offset(field) XtOffset(GasWidget, gas.field)
    /* {name, class, type, size, offset, default_type, default_addr}, */
/*    { XtNgasResource, XtCGasResource, XtRGasResource, sizeof(char*),*/
/*	  offset(resource), XtRString, "default" },*/
    { XtNresize, XtCCallback, XtRCallback, sizeof(char*),
	  offset(resource), XtRPointer, NULL },
#undef offset
};

static void GasAction(/* Widget, XEvent*, String*, Cardinal* */);
static void Resize();

static XtActionsRec actions[] =
{
  /* {name, procedure}, */
    {"gas",	GasAction},
};

static char translations[] =
"<Key>:		gas()	\n\
";

GasClassRec gasClassRec = {
  { /* core fields */
    /* superclass		*/	(WidgetClass) &widgetClassRec,
    /* class_name		*/	"Gas",
    /* widget_size		*/	sizeof(GasRec),
    /* class_initialize		*/	NULL,
    /* class_part_initialize	*/	NULL,
    /* class_inited		*/	FALSE,
    /* initialize		*/	NULL,
    /* initialize_hook		*/	NULL,
    /* realize			*/	XtInheritRealize,
    /* actions			*/	actions,
    /* num_actions		*/	XtNumber(actions),
    /* resources		*/	resources,
    /* num_resources		*/	XtNumber(resources),
    /* xrm_class		*/	NULLQUARK,
    /* compress_motion		*/	TRUE,
    /* compress_exposure	*/	TRUE,
    /* compress_enterleave	*/	TRUE,
    /* visible_interest		*/	FALSE,
    /* destroy			*/	NULL,
    /* resize			*/	Resize,
    /* expose			*/	NULL,
    /* set_values		*/	NULL,
    /* set_values_hook		*/	NULL,
    /* set_values_almost	*/	XtInheritSetValuesAlmost,
    /* get_values_hook		*/	NULL,
    /* accept_focus		*/	NULL,
    /* version			*/	XtVersion,
    /* callback_private		*/	NULL,
    /* tm_table			*/	translations,
    /* query_geometry		*/	XtInheritQueryGeometry,
    /* display_accelerator	*/	XtInheritDisplayAccelerator,
    /* extension		*/	NULL
  },
  { /* gas fields */
    /* empty			*/	0
  }
};

WidgetClass gasWidgetClass = (WidgetClass)&gasClassRec;

static void
GasAction(w, event, params, num_params)			/* ARGSUSED */
     Widget w;
     XEvent *event;
     String *params;		/* unused */
     Cardinal *num_params;	/* unused */
{
  XtCallCallbacks(w, XtNcallback, (caddr_t)event);
}

static void
Resize(w)
Widget w;
{
  XtCallCallbacks(w, XtNresize, NULL);
}
